// @mysten/zksend@0.8.0/dist/cjs/links/builder.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __accessCheck = (obj, member, msg) => {
  if (!member.has(obj))
    throw TypeError("Cannot " + msg);
};
var __privateGet = (obj, member, getter) => {
  __accessCheck(obj, member, "read from private field");
  return getter ? getter.call(obj) : member.get(obj);
};
var __privateAdd = (obj, member, value) => {
  if (member.has(obj))
    throw TypeError("Cannot add the same private member more than once");
  member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
};
var __privateSet = (obj, member, value, setter) => {
  __accessCheck(obj, member, "write to private field");
  setter ? setter.call(obj, value) : member.set(obj, value);
  return value;
};
var __privateMethod = (obj, member, method) => {
  __accessCheck(obj, member, "access private method");
  return method;
};
var builder_exports = {};
__export(builder_exports, {
  ZkSendLinkBuilder: () => ZkSendLinkBuilder
});
module.exports = __toCommonJS(builder_exports);
var import_client = require("@mysten/sui.js/client");
var import_cryptography = require("@mysten/sui.js/cryptography");
var import_ed25519 = require("@mysten/sui.js/keypairs/ed25519");
var import_transactions = require("@mysten/sui.js/transactions");
var import_utils = require("@mysten/sui.js/utils");
var import_zk_bag = require("./zk-bag.js");
var _host, _path, _client, _redirect, _coinsByType, _contract, _objectsToTransfer, objectsToTransfer_fn, _createSendTransactionWithoutContract, createSendTransactionWithoutContract_fn, _estimateClaimGasFee, estimateClaimGasFee_fn, _getCoinsByType, getCoinsByType_fn;
const DEFAULT_ZK_SEND_LINK_OPTIONS = {
  host: "https://zksend.com",
  path: "/claim",
  network: "mainnet"
};
const SUI_COIN_TYPE = (0, import_utils.normalizeStructTag)(import_utils.SUI_TYPE_ARG);
const _ZkSendLinkBuilder = class {
  constructor({
    host = DEFAULT_ZK_SEND_LINK_OPTIONS.host,
    path = DEFAULT_ZK_SEND_LINK_OPTIONS.path,
    keypair = new import_ed25519.Ed25519Keypair(),
    network = DEFAULT_ZK_SEND_LINK_OPTIONS.network,
    client = new import_client.SuiClient({ url: (0, import_client.getFullnodeUrl)(network) }),
    sender,
    redirect,
    contract = network === "mainnet" ? import_zk_bag.MAINNET_CONTRACT_IDS : void 0
  }) {
    __privateAdd(this, _objectsToTransfer);
    __privateAdd(this, _createSendTransactionWithoutContract);
    __privateAdd(this, _estimateClaimGasFee);
    __privateAdd(this, _getCoinsByType);
    this.objectIds = /* @__PURE__ */ new Set();
    this.objectRefs = [];
    this.balances = /* @__PURE__ */ new Map();
    __privateAdd(this, _host, void 0);
    __privateAdd(this, _path, void 0);
    __privateAdd(this, _client, void 0);
    __privateAdd(this, _redirect, void 0);
    __privateAdd(this, _coinsByType, /* @__PURE__ */ new Map());
    __privateAdd(this, _contract, void 0);
    __privateSet(this, _host, host);
    __privateSet(this, _path, path);
    __privateSet(this, _redirect, redirect);
    this.keypair = keypair;
    __privateSet(this, _client, client);
    this.sender = (0, import_utils.normalizeSuiAddress)(sender);
    if (contract) {
      __privateSet(this, _contract, new import_zk_bag.ZkBag(contract.packageId, contract));
    }
  }
  addClaimableMist(amount) {
    this.addClaimableBalance(SUI_COIN_TYPE, amount);
  }
  addClaimableBalance(coinType, amount) {
    const normalizedType = (0, import_utils.normalizeStructTag)(coinType);
    this.balances.set(normalizedType, (this.balances.get(normalizedType) ?? 0n) + amount);
  }
  addClaimableObject(id) {
    this.objectIds.add(id);
  }
  addClaimableObjectRef(ref, type) {
    this.objectRefs.push({ ref, type });
  }
  getLink() {
    const link = new URL(__privateGet(this, _host));
    link.pathname = __privateGet(this, _path);
    link.hash = `${__privateGet(this, _contract) ? "$" : ""}${(0, import_utils.toB64)(
      (0, import_cryptography.decodeSuiPrivateKey)(this.keypair.getSecretKey()).secretKey
    )}`;
    if (__privateGet(this, _redirect)) {
      link.searchParams.set("redirect_url", __privateGet(this, _redirect).url);
      if (__privateGet(this, _redirect).name) {
        link.searchParams.set("name", __privateGet(this, _redirect).name);
      }
    }
    return link.toString();
  }
  async create({
    signer,
    ...options
  }) {
    const txb = await this.createSendTransaction(options);
    const result = await __privateGet(this, _client).signAndExecuteTransactionBlock({
      transactionBlock: await txb.build({ client: __privateGet(this, _client) }),
      signer
    });
    if (options.waitForTransactionBlock) {
      await __privateGet(this, _client).waitForTransactionBlock({ digest: result.digest });
    }
    return result;
  }
  async createSendTransaction({
    transactionBlock = new import_transactions.TransactionBlock(),
    calculateGas
  } = {}) {
    if (!__privateGet(this, _contract)) {
      return __privateMethod(this, _createSendTransactionWithoutContract, createSendTransactionWithoutContract_fn).call(this, { transactionBlock, calculateGas });
    }
    transactionBlock.setSenderIfNotSet(this.sender);
    return _ZkSendLinkBuilder.createLinks({
      transactionBlock,
      client: __privateGet(this, _client),
      contract: __privateGet(this, _contract).ids,
      links: [this]
    });
  }
  async createSendToAddressTransaction({
    transactionBlock = new import_transactions.TransactionBlock(),
    address
  }) {
    const objectsToTransfer = (await __privateMethod(this, _objectsToTransfer, objectsToTransfer_fn).call(this, transactionBlock)).map(
      (obj) => obj.ref
    );
    transactionBlock.setSenderIfNotSet(this.sender);
    transactionBlock.transferObjects(objectsToTransfer, address);
    return transactionBlock;
  }
  static async createLinks({
    links,
    network = "mainnet",
    client = new import_client.SuiClient({ url: (0, import_client.getFullnodeUrl)(network) }),
    transactionBlock = new import_transactions.TransactionBlock(),
    contract: contractIds = import_zk_bag.MAINNET_CONTRACT_IDS
  }) {
    const contract = new import_zk_bag.ZkBag(contractIds.packageId, contractIds);
    const store = transactionBlock.object(contract.ids.bagStoreId);
    const coinsByType = /* @__PURE__ */ new Map();
    const allIds = links.flatMap((link) => [...link.objectIds]);
    const sender = links[0].sender;
    transactionBlock.setSenderIfNotSet(sender);
    await Promise.all(
      [...new Set(links.flatMap((link) => [...link.balances.keys()]))].map(async (coinType) => {
        const coins = await client.getCoins({
          coinType,
          owner: sender
        });
        coinsByType.set(
          coinType,
          coins.data.filter((coin) => !allIds.includes(coin.coinObjectId))
        );
      })
    );
    const objectRefs = /* @__PURE__ */ new Map();
    const pageSize = 50;
    let offset = 0;
    while (offset < allIds.length) {
      let chunk = allIds.slice(offset, offset + pageSize);
      offset += pageSize;
      const objects = await client.multiGetObjects({
        ids: chunk,
        options: {
          showType: true
        }
      });
      for (const [i, res] of objects.entries()) {
        if (!res.data || res.error) {
          throw new Error(`Failed to load object ${chunk[i]} (${res.error?.code})`);
        }
        objectRefs.set(chunk[i], {
          ref: transactionBlock.objectRef({
            version: res.data.version,
            digest: res.data.digest,
            objectId: res.data.objectId
          }),
          type: res.data.type
        });
      }
    }
    const mergedCoins = /* @__PURE__ */ new Map([
      [SUI_COIN_TYPE, transactionBlock.gas]
    ]);
    for (const [coinType, coins] of coinsByType) {
      if (coinType === SUI_COIN_TYPE) {
        continue;
      }
      const [first, ...rest] = coins.map(
        (coin) => transactionBlock.objectRef({
          objectId: coin.coinObjectId,
          version: coin.version,
          digest: coin.digest
        })
      );
      if (rest.length > 0) {
        transactionBlock.mergeCoins(first, rest);
      }
      mergedCoins.set(coinType, transactionBlock.object(first));
    }
    for (const link of links) {
      const receiver = link.keypair.toSuiAddress();
      contract.new(transactionBlock, { arguments: [store, receiver] });
      link.objectRefs.forEach(({ ref, type }) => {
        contract.add(transactionBlock, {
          arguments: [store, receiver, ref],
          typeArguments: [type]
        });
      });
      link.objectIds.forEach((id) => {
        const object = objectRefs.get(id);
        if (!object) {
          throw new Error(`Object ${id} not found`);
        }
        contract.add(transactionBlock, {
          arguments: [store, receiver, object.ref],
          typeArguments: [object.type]
        });
      });
    }
    for (const [coinType, merged] of mergedCoins) {
      const linksWithCoin = links.filter((link) => link.balances.has(coinType));
      if (linksWithCoin.length === 0) {
        continue;
      }
      const balances = linksWithCoin.map((link) => link.balances.get(coinType));
      const splits = transactionBlock.splitCoins(merged, balances);
      for (const [i, link] of linksWithCoin.entries()) {
        contract.add(transactionBlock, {
          arguments: [store, link.keypair.toSuiAddress(), splits[i]],
          typeArguments: [`0x2::coin::Coin<${coinType}>`]
        });
      }
    }
    return transactionBlock;
  }
};
let ZkSendLinkBuilder = _ZkSendLinkBuilder;
_host = new WeakMap();
_path = new WeakMap();
_client = new WeakMap();
_redirect = new WeakMap();
_coinsByType = new WeakMap();
_contract = new WeakMap();
_objectsToTransfer = new WeakSet();
objectsToTransfer_fn = async function(txb) {
  const objectIDs = [...this.objectIds];
  const refsWithType = this.objectRefs.concat(
    (objectIDs.length > 0 ? await __privateGet(this, _client).multiGetObjects({
      ids: objectIDs,
      options: {
        showType: true
      }
    }) : []).map((res, i) => {
      if (!res.data || res.error) {
        throw new Error(`Failed to load object ${objectIDs[i]} (${res.error?.code})`);
      }
      return {
        ref: txb.objectRef({
          version: res.data.version,
          digest: res.data.digest,
          objectId: res.data.objectId
        }),
        type: res.data.type
      };
    })
  );
  for (const [coinType, amount] of this.balances) {
    if (coinType === SUI_COIN_TYPE) {
      const [sui] = txb.splitCoins(txb.gas, [amount]);
      refsWithType.push({
        ref: sui,
        type: `0x2::coin::Coin<${coinType}>`
      });
    } else {
      const coins = (await __privateMethod(this, _getCoinsByType, getCoinsByType_fn).call(this, coinType)).map((coin) => coin.coinObjectId);
      if (coins.length > 1) {
        txb.mergeCoins(coins[0], coins.slice(1));
      }
      const [split] = txb.splitCoins(coins[0], [amount]);
      refsWithType.push({
        ref: split,
        type: `0x2::coin::Coin<${coinType}>`
      });
    }
  }
  return refsWithType;
};
_createSendTransactionWithoutContract = new WeakSet();
createSendTransactionWithoutContract_fn = async function({
  transactionBlock: txb = new import_transactions.TransactionBlock(),
  calculateGas
} = {}) {
  const gasEstimateFromDryRun = await __privateMethod(this, _estimateClaimGasFee, estimateClaimGasFee_fn).call(this);
  const baseGasAmount = calculateGas ? await calculateGas({
    balances: this.balances,
    objects: [...this.objectIds],
    gasEstimateFromDryRun
  }) : gasEstimateFromDryRun * 2n;
  const gasWithBuffer = baseGasAmount + 1013n;
  const roundedGasAmount = gasWithBuffer - gasWithBuffer % 1000n - 13n;
  const address = this.keypair.toSuiAddress();
  const objectsToTransfer = (await __privateMethod(this, _objectsToTransfer, objectsToTransfer_fn).call(this, txb)).map((obj) => obj.ref);
  const [gas] = txb.splitCoins(txb.gas, [roundedGasAmount]);
  objectsToTransfer.push(gas);
  txb.setSenderIfNotSet(this.sender);
  txb.transferObjects(objectsToTransfer, address);
  return txb;
};
_estimateClaimGasFee = new WeakSet();
estimateClaimGasFee_fn = async function() {
  const txb = new import_transactions.TransactionBlock();
  txb.setSender(this.sender);
  txb.setGasPayment([]);
  txb.transferObjects([txb.gas], this.keypair.toSuiAddress());
  const idsToTransfer = [...this.objectIds];
  for (const [coinType] of this.balances) {
    const coins = await __privateMethod(this, _getCoinsByType, getCoinsByType_fn).call(this, coinType);
    if (!coins.length) {
      throw new Error(`Sending account does not contain any coins of type ${coinType}`);
    }
    idsToTransfer.push(coins[0].coinObjectId);
  }
  if (idsToTransfer.length > 0) {
    txb.transferObjects(
      idsToTransfer.map((id) => txb.object(id)),
      this.keypair.toSuiAddress()
    );
  }
  const result = await __privateGet(this, _client).dryRunTransactionBlock({
    transactionBlock: await txb.build({ client: __privateGet(this, _client) })
  });
  return BigInt(result.effects.gasUsed.computationCost) + BigInt(result.effects.gasUsed.storageCost) - BigInt(result.effects.gasUsed.storageRebate);
};
_getCoinsByType = new WeakSet();
getCoinsByType_fn = async function(coinType) {
  if (__privateGet(this, _coinsByType).has(coinType)) {
    return __privateGet(this, _coinsByType).get(coinType);
  }
  const coins = await __privateGet(this, _client).getCoins({
    coinType,
    owner: this.sender
  });
  __privateGet(this, _coinsByType).set(coinType, coins.data);
  return coins.data;
};
